pkg/collector/corechecks/gpu/gpu.go

```diff
@@ -0,0 +1,206 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```

Comment:
i suggest taking a look at servicediscovery core check and implementing the setup similarly.

- we want to bail out fast for non linux environments
- it could be useful to have an interface and implement a mock (for testing)
- we should try to initialize the sysproberemoteclient at the very beginning as it can be a long operation (in case the unix socket is unavailable for some reason)

Reply:
> we want to bail out fast for non linux environments

The check only builds in Linux targets, in non-Linux environments it won't exist (see `gpu_stub.go`)

> it could be useful to have an interface and implement a mock (for testing)

Most of the testable logic is in separate components, the check is mostly plumbing 

> we should try to initialize the sysproberemoteclient at the very beginning as it can be a long operation (in case the unix socket is unavailable for some reason)

[There was a comment about this](https://github.com/DataDog/datadog-agent/pull/29647#discussion_r1790191602), apparently it seems it's better to try to connect inside of the check run itself. In any case, if the socket isn't available there's nothing else we can do but wait, I think once it's available with system-probe listening the setup itself is fast.

Reply:
_OS flavor:_ Nice one! TIL

_testing:_ so if i want to test the check logic, can i do it without the systemprobe component? i think ensureInitialized will return an error in that case. 

_SystemProbeUtil:_ what about the race condition we talked in the past? if ensureInitialized is called from Run, what happens if next check iteration is executed before the first current one finished?

Reply:
_testing_: no, but I think in that case it would be easier to extract the code that processes the data to a function that receives the `model.GPUStats` object and the `Sender`, which would probably be simpler (we wouldn't have to mock just the sysprobeutil but also the NVML library).

_ SystemProbeUtil_: [based on this code](https://github.com/DataDog/datadog-agent/blob/guillermo.julian/cuda-metrics-agent/comp/collector/collector/collectorimpl/internal/middleware/check_wrapper.go#L44) I think that only one `Run()` call can be done concurrently, as it's wrapped in a lock.

Reply:
created ticket - https://datadoghq.atlassian.net/browse/EBPF-578 for a separate PR.

I will test the usage of the check middleware (that ensures the lock is taken when executing the Run method)

Reply:
Tested the call stack, the check middleware is running so `Run` is always running under a lock.